python/pyarrow/parquet/core.py

```diff
@@ -0,0 +1,3467 @@
+
+    >>> pq.write_metadata(
+    ...     table.schema, 'dataset_metadata/_metadata',
+    ...     metadata_collector=metadata_collector)
+    """
+    writer = ParquetWriter(where, schema, **kwargs)
+    writer.close()
+
+    if metadata_collector is not None:
+        # ParquetWriter doesn't expose the metadata until it's written. Write
+        # it and read it again.
+        metadata = read_metadata(where)
+        for m in metadata_collector:
+            metadata.append_row_groups(m)
+        metadata.write_metadata_file(where)
+
+
+def read_metadata(where, memory_map=False, decryption_properties=None):
+    """
+    Read FileMetaData from footer of a single Parquet file.
+
+    Parameters
+    ----------
+    where : str (file path) or file-like object
+    memory_map : bool, default False
+        Create memory map when the source is a file path.
+    decryption_properties : FileDecryptionProperties, default None
+        Decryption properties for reading encrypted Parquet files.
+
+    Returns
+    -------
+    metadata : FileMetaData
+
+    Examples
+    --------
+    >>> import pyarrow as pa
+    >>> import pyarrow.parquet as pq
+    >>> table = pa.table({'n_legs': [4, 5, 100],
+    ...                   'animal': ["Dog", "Brittle stars", "Centipede"]})
+    >>> pq.write_table(table, 'example.parquet')
+
+    >>> pq.read_metadata('example.parquet')
+    <pyarrow._parquet.FileMetaData object at ...>
+      created_by: parquet-cpp-arrow version ...
+      num_columns: 2
+      num_rows: 3
+      num_row_groups: 1
+      format_version: 2.6
+      serialized_size: ...
+    """
+    return ParquetFile(where, memory_map=memory_map,
+                       decryption_properties=decryption_properties).metadata
+
+
+def read_schema(where, memory_map=False, decryption_properties=None):
+    """
+    Read effective Arrow schema from Parquet file metadata.
+
+    Parameters
+    ----------
+    where : str (file path) or file-like object
+    memory_map : bool, default False
+        Create memory map when the source is a file path.
+    decryption_properties : FileDecryptionProperties, default None
+        Decryption properties for reading encrypted Parquet files.
+
+    Returns
+    -------
+    schema : pyarrow.Schema
+
+    Examples
+    --------
+    >>> import pyarrow as pa
+    >>> import pyarrow.parquet as pq
+    >>> table = pa.table({'n_legs': [4, 5, 100],
+    ...                   'animal': ["Dog", "Brittle stars", "Centipede"]})
+    >>> pq.write_table(table, 'example.parquet')
+
+    >>> pq.read_schema('example.parquet')
+    n_legs: int64
+    animal: string
+    """
+    return ParquetFile(
+        where, memory_map=memory_map,
+        decryption_properties=decryption_properties).schema.to_arrow_schema()
+
+
+# re-export everything
+# std `from . import *` ignores symbols with leading `_`
+__all__ = [
+    m for m, _ in inspect.getmembers(sys.modules[__name__])
+]
```

Comment:
It seems this is the same as the following, but much slower?
```suggestion
__all__ = list(__dict__)
```

Reply:
Maybe I'm missing somehting, but I get 
```
-> 3463 __all__ = list(__dict__)

NameError: name '__dict__' is not defined
```

Changing it to `list(sys.modules[__name__].__dict__)` and a smidgen faster. Thanks!

```python
In [10]: N = 1_000

In [11]: # __dict__

In [12]: timeit.timeit('_ = importlib.reload(core)', number=N, globals=globals()) / N
Out[12]: 0.0005946487420005723

In [13]: # inspect.getmembers

In [14]: timeit.timeit('_ = importlib.reload(core)', number=N, globals=globals()) / N
Out[14]: 0.000692891882998083

In [15]: 
```

Reply:
By the way, since you're using IPython, you probably know it but the `%timeit` macro is useful :-)

Reply:
Recently switched to it actually, thanks for the tip! 👍